### ISSUE TITLE
No Need to compute Generator Gradients when training Discriminator

### ISSUE THREAD

**Issue by Skylion007** [akanimax/MSG-GAN#4]
Since Discriminator and Generator are trained separately, we don't need the gradients of the generator when training the discriminator.

Here are some relevant issues:
https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/d7e7ed851b3e5ac01167197011bf8204015a1ce7/models/cycle_gan_model.py#L140
https://github.com/pytorch/pytorch/issues/5349

Fixing this will probably speed up training.

**Comment by akanimax** [comment]
Thanks **Aaron** for sending the links. They are very helpful. I got the idea that you are suggesting. I hope this will also reduce the memory consumption apart from speeding up training. :+1: 

Cheers :beers:!

**Comment by Skylion007** [comment]
On second thought, I didn't catch the .detach() operation which seems like it should stop the backprop of gradients. Some of the code I was looking it was older PyTorch code. Sorry about that.

**Comment by akanimax** [comment]
I am sure `.detach()` stops the operations from performing the backpropogation, but I think it still holds the grad buffers in the tensors which might consume memory. I will still take a look at that.
